### PR TITLE
Docker container names

### PR DESCRIPTION
**Pull request by gg7**
Works on my machine ™

### CODE REVIEW

File: cgroup_top.py [hunk 1]
```diff
@@ -49,6 +50,38 @@ def cmd_exists(cmd):
     except OSError:
         return False
 
+
+def docker_container_name(container_id, default, cache=dict()):
+    cached_name = cache.get(container_id)
+    if cached_name:
+        return cached_name
+
+    sp = subprocess.Popen(['docker', 'inspect', container_id], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
```
Comment (yadutaf) [inline cgroup_top.py]: If `docker` binary is not in the PATH, it will raise `OSError: [Errno 2] No such file or directory`

File: cgroup_top.py [hunk 2]
```diff
@@ -49,6 +50,38 @@ def cmd_exists(cmd):
     except OSError:
         return False
 
+
+def docker_container_name(container_id, default, cache=dict()):
```
Comment (yadutaf) [inline cgroup_top.py]: This is a clever way to do caching but few people know that the "default" value is shared between calls. Can you add a comment to hint them?

File: cgroup_top.py [hunk 3]
```diff
@@ -49,6 +50,38 @@ def cmd_exists(cmd):
...
+
+    if sp.returncode is None:
+        try:
+            sp.kill()
+        except OSError:
+            # OSError: [Errno 3] No such process
+            pass
+    elif sp.returncode == 0:
+        stdout, _stderr = sp.communicate()
+        data = json.loads(stdout)
```
Comment (yadutaf) [inline cgroup_top.py]: What if output is garbage ?

File: cgroup_top.py [hunk 4]
```diff
@@ -190,10 +223,21 @@ def __init__(self, path, base_path):
         self.base_path = base_path
 
     @property
-    def name(self):
+    def short_path(self):
         return self.path[len(self.base_path):] or '/'
 
     @property
+    def name(self):
+        if self.type == 'docker':
```
Comment (yadutaf) [inline cgroup_top.py]: may be add here `and HAS_DOCKER` or something similar. People tend to use ctop from a container with only access to the cgroups but no access to the docker binary or socket.

**Comment by yadutaf** [comment]
Hi @gg7, thanks for your contribution ! Here is a quick review.

**Comment by gg7** [comment]
Good points @yadutaf.

File: cgroup_top.py [hunk 5]
```diff
@@ -100,6 +103,60 @@ def cmd_exists(cmd):
...
+    # pointlessly retrying to get the container name if something goes wrong.
+    cached_name = cache.get(container_id)
+    if cached_name:
+        return cached_name
+
+    try:
+        sp = subprocess.Popen(['docker', 'inspect', container_id],
+                              stdout=subprocess.PIPE,
+                              stderr=subprocess.PIPE)
+    except OSError:
```
Comment (yadutaf) [inline cgroup_top.py:126]: Actually, with the `HAS_DOCKER` check lower, this should never trigger. But that's harmless / future proof.

File: cgroup_top.py [hunk 6]
```diff
@@ -100,6 +103,60 @@ def cmd_exists(cmd):
...
 
+
+def strip_prefix(prefix, text):
+    if text.startswith(prefix):
+        return text[len(prefix):]
+    return text
+
+
+def docker_container_name(container_id, default, cache=dict()):
+    # Python’s default arguments are evaluated when the function is
```
Comment (yadutaf) [inline cgroup_top.py]: Funny UTF-8 sequence here :) Can you use regular `'` in "Python's" ? In vim it displays as a nice `# Python�~@~Ys`. Other ones are OK.

File: cgroup_top.py (hunk 5 above)
Comment (gg7) [inline cgroup_top.py:126]: Maybe an evil administrator executed `rm $(which docker)` :)

File: cgroup_top.py (hunk 6 above)
Comment (gg7) [inline cgroup_top.py]: Fixed. I use vim as well and it displays it correctly though.

**Comment by gg7** [comment]
Feel free to squash / cherry pick my commits.

File: cgroup_top.py (hunk 5 above)
Comment (yadutaf) [inline cgroup_top.py:126]: Granted :)

**Comment by yadutaf** [comment]
Merged, thx ! (I never squash other people commits btw)

**Pull request merged**